src/arviz_stats/base/core.py

```diff
@@ -252,40 +252,94 @@ def _hdi_nearest(self, ary, prob, circular, skipna):
 
         return hdi_interval
 
-    def _hdi_multimodal(self, ary, prob, skipna, max_modes):
+    def _hdi_multimodal_continuous(self, ary, prob, skipna, max_modes, circular, **kwargs):
         """Compute HDI if the distribution is multimodal."""
         ary = ary.flatten()
         if skipna:
             ary = ary[~np.isnan(ary)]
 
-        if ary.dtype.kind == "f":
-            bins, density, _ = self.kde(ary)
-            lower, upper = bins[0], bins[-1]
-            range_x = upper - lower
-            dx = range_x / len(density)
+        bins, density, _ = self.kde(ary, circular=circular, **kwargs)
+        lower, upper = bins[0], bins[-1]
+        range_x = upper - lower
+        dx = range_x / len(density)
+
+        hdi_intervals, interval_probs = self._hdi_from_bin_probabilities(
+            bins, density, prob, circular, dx
+        )
+
+        return self._pad_hdi_to_maxmodes(hdi_intervals, interval_probs, max_modes)
+
+    def _hdi_multimodal_discrete(self, ary, prob, bins, max_modes):
+        """Compute HDI if the distribution is multimodal."""
+        ary = ary.flatten()
+
+        if bins is None:
+            bins, density = np.unique(ary, return_counts=True)
+            dx = 1
         else:
             bins = self._get_bins(ary)
-            density, _ = self._histogram(ary, bins=bins, density=True)
+            density, _ = self._histogram(ary, bins=bins)
             dx = np.diff(bins)[0]
 
-        density *= dx
-
-        idx = np.argsort(-density)
-        intervals = bins[idx][density[idx].cumsum() <= prob]
-        intervals.sort()
+        hdi_intervals, interval_probs = self._hdi_from_bin_probabilities(
+            bins, density, prob, False, dx
+        )
 
-        intervals_splitted = np.split(intervals, np.where(np.diff(intervals) >= dx * 1.1)[0] + 1)
+        return self._pad_hdi_to_maxmodes(hdi_intervals, interval_probs, max_modes)
 
-        hdi_intervals = np.full((max_modes, 2), np.nan)
-        for i, interval in enumerate(intervals_splitted):
-            if i == max_modes:
-                warnings.warn(
-                    f"found more modes than {max_modes}, returning only the first {max_modes} modes"
-                )
-                break
-            if interval.size == 0:
-                hdi_intervals[i] = np.asarray([bins[0], bins[0]])
-            else:
-                hdi_intervals[i] = np.asarray([interval[0], interval[-1]])
+    def _hdi_from_bin_probabilities(self, bins, bin_probs, prob, circular, dx):  # pylint: disable=no-self-use
+        if circular:
+            # standardize bins to [-pi, pi]
+            bins = np.fmod(bins + np.pi, 2 * np.pi) - np.pi
+            sorted_idx = np.argsort(bins)
+            bins = bins[sorted_idx]
+            bin_probs = bin_probs[sorted_idx]
+
+        # find idx of bins in the interval
+        sorted_idx = np.argsort(bin_probs)[::-1]
+        bin_probs = bin_probs / bin_probs.sum()
```

Comment:
They're not exactly normalized, or at least not in the way that we need for this implementation:
```python
>>> import numpy as np
>>> from arviz_base import load_arviz_data
>>> import arviz_stats
>>> 
>>> data = load_arviz_data("centered_eight")
>>> kde = data.posterior.theta.azstats.kde()
>>> dx = kde.sel(plot_axis='x').diff(dim='kde_dim').sel(kde_dim=0)
>>> kde.sel(plot_axis='y').sum(dim='kde_dim') * dx
<xarray.DataArray 'theta' (school: 8)> Size: 64B
array([0.99830641, 0.99826941, 0.9979747 , 0.9978814 , 0.99818729,
       0.99800671, 0.9979463 , 0.9979312 ])
Coordinates:
  * school    (school) <U16 512B 'Choate' 'Deerfield' ... 'Mt. Hermon'
    bw_theta  (school) float64 64B 0.5168 0.5102 0.5383 ... 0.5142 0.4794 0.5383
```

Reply:
We should definitely fix this (even if we kept the normalization here). I'll try to take a look and see why this is happening, maybe we are applying the theoretical factor to get the normalized kde but floats have other plans.

Reply:
I think it's the choice of `kernel_n`. e.g. we can use the alternative `kernel_n = 2 * math.ceil(4 * bw) + 1`, which effectively chooses the smallest integer so that all kernel evaluations fall within 4 standard deviations of the mode, and the mode is included. The result is much better normalized:
```python
>>> data = load_arviz_data("centered_eight")
>>> kde = data.posterior.theta.azstats.kde()
>>> dx = kde.sel(plot_axis='x').diff(dim='kde_dim').sel(kde_dim=0)
>>> kde.sel(plot_axis='y').sum(dim='kde_dim') * dx
<xarray.DataArray 'theta' (school: 8)> Size: 64B
array([0.99998307, 0.99997361, 0.99997543, 0.99997636, 0.9999717 ,
       0.99997684, 0.99997628, 0.99997567])
Coordinates:
  * school    (school) <U16 512B 'Choate' 'Deerfield' ... 'Mt. Hermon'
    bw_theta  (school) float64 64B 0.5168 0.5102 0.5383 ... 0.5142 0.4794 0.5383
```

Going to 5 standard deviations gives us:
```python
>>> data = load_arviz_data("centered_eight")
>>> kde = data.posterior.theta.azstats.kde()
>>> dx = kde.sel(plot_axis='x').diff(dim='kde_dim').sel(kde_dim=0)
>>> kde.sel(plot_axis='y').sum(dim='kde_dim') * dx
<xarray.DataArray 'theta' (school: 8)> Size: 64B
array([0.99999983, 0.9999998 , 0.99999985, 0.99999974, 0.99999977,
       0.99999986, 0.99999974, 0.99999972])
Coordinates:
  * school    (school) <U16 512B 'Choate' 'Deerfield' ... 'Mt. Hermon'
    bw_theta  (school) float64 64B 0.5168 0.5102 0.5383 ... 0.5142 0.4794 0.5383
```

Since we hardcode that the convolution must be direct, I think we pay a performance cost by increasing `kernel_n`. The convolution could also entirely be done in frequency space if we perform an RFFT of the histogram ourselves and compute the characteristic function of the kernel directly. The bottleneck is then the O(n*log(n)) fft, not the O(n) kernel computation, so we wouldn't need or want to truncate at all. There's some additional overhead to performing an FFT that we would pay, but for ISJ bandwidth selection we already compute the FFT of the histogram to compute the DCT, so in principle with a refactor we could eliminate this overhead.

I think the simplest solution for now is to adopt one of the `kernel_n` options above. FWIW, Wand in https://doi.org/10.1201/b14876 recommends 4 standard deviations. With this choice, we could always explicitly renormalize just to be a little more precise.

Reply:
sounds good, yes